lib/posix/close.h

```diff
@@ -0,0 +1,20 @@
+#ifndef LIB_POSIX_CLOSE_H_
+#define LIB_POSIX_CLOSE_H_
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "lib/error/status.h"
+#include "lib/posix/errno.h"
+#include "lib/posix/file_descriptor.h"
+
+namespace posix {
+
+inline ::error::Status Close(const FileDescriptor fd) {
+  const auto rv = ::close(GetValue(fd));
+  return OkStatusOrCaptureErrnoIf(-1 == rv, "close() failed");
+}
```

Comment:
When do we transition into the OO world? I saw the use of raw IsOk() and IsError() functions for dealing with Status elsewhere and thought hmmm, ok... But, do we want a File abstraction at some point with Close() as a member?

Then, if we have a Close() that returns Status, don't we also want an Open() that returns some kind of Result based on std::variant? The examples show the result of open passed directly to FileDescriptor's constructor, which can very easily result in invalid FileDescriptor instances. Shouldn't we catch this earlier?

Also, shouldn't Close() also test IsWellFormed() and return an error Status when not? And, might this not all go away if your interface to Open() can only ever return well formed FileDescriptors in the first place?

Reply:
Great question!

I'm intentionally avoiding certain OOP patterns, in OOP there is a tendency for certain types of objects to grow many unrelated member functions. In particular, different users may want to provide utility functions that are only relevant to a subset of code and may even be mutually incompatible with equivalent utilities in a different section of a code base. When using member functions there is no way to place different subsets of functionality in namespaces except for inheritance which unfortunately comes with the difficult decisions around multiple inheritance when trying to compose functionality.

Here's an example of what IsOk() provides that .Ok() cannot: Currently RETURN_IF_ERROR() works with util::Status, however in the future we may interact with objects from a third party library that we are unable to extend. If RETURN_IF_ERROR() depends on .Ok(), I don't have a good way to make new types work with RETURN_IF_ERROR(). However with IsOk(), all I need to do is implement IsOk() in an appropriate namespace. Overloads like this (typeclasses) are the primary form of inheritance in Haskell - C++ is moving in a similar direction with Concepts.

In the case of FileDescriptor, functionality of file descriptors is broken into ownership (UniqueFileDescriptor) and "a value that should be interpreted as a file descriptor" (FileDescriptor). When a function accepts a FD as input but does not take ownership a value of type FileDescriptor should be used. When ownership is held or passed then UniqueFileDescriptor is used with std::move.

Why not create a class with Open() and Close() member functions? POSIX file descriptors are a very general concept - they're used by many subsystems beyond files and open. UniqueFileDescriptor captures the concept of "an OS handle that must have close() called on it exactly once", but beyond this there is not much that they have in common.

In the case of eBPF, the bpf() syscall returns a file descriptor that could represent a loaded program or a map. open(), socket(), accept(), pipe2(), and a myriad of ioctls all return a file descriptor on success.

You could argue that read() and write() are universal, however send(), recv(), sendmsg() and recvmsg() are used on stream sockets and datagram sockets. Some file descriptors support mmap(), some fcntl() and others setsockopt(). There's not a great way to compose these subsets using OOP inheritance.

Your comment about open() in my example potentially placing an error return value in UniqueFileDescriptor is a great point. I purposely glossed over that issue since it isn't necessary to understand the auto-close concept and could distract from the core idea. I can elaborate if you think it is too confusing. In practice each function that can return a file descriptor will be wrapped in the following way:

error::StatusOr<posix::UniqueFileDescriptor> Open(const std::string& path, const int flags);

...where the Open() function is responsible for checking the return value of the open() syscall and returning either a non-ok Status or a owned file descriptor.

A syscall that that takes a file descriptors as input might look like this:
error::StatusOr<size_t> Read(const posix::FileDescriptor fd, void* const dst, const size_t count);

Hopefully once a few examples using the API exist it should make more sense. The primary benefit is that different kernel API wrappers can share the file descriptor infrastructure without polluting a common namespace or causing churn in a core class. When you consider that there are order of a hundred syscalls that operate on file descriptors and even more ioctls, it should be clear member functions and inheritance do not model this well.

Reply:
Thank you for the detailed rationale and I'm happy to wait for the safer Open() functionality if you consider that a separate body of work.

I think the comment about Close() testing for well formed descriptors should still be addressed?

Reply:
After your comments, I propose that we change UniqueFileDescriptor's implementation to look like this:

struct UniqueFileDescriptorDtor {
  void operator()(const FileDescriptor fd) {
    INVARIANT_T(IsWellFormed(fd));
    // The OS sometimes returns an information error, particularly in the case of stream sockets.
    // The contract provided by the close() syscall is that the file descriptor must be closed by 
    // the time close() returns because there is no way for a user to recover from an error that
    // leaves the FD in an indeterminate state.
    // Nothing we can do about an error here.
    base::Ignore(Close(fd));
  }
};

I'll argue that syscall wrappers taking FileDescriptor as input should never check for the validity of the input file descriptor (this is too many checks) and we should instead have the invariant that UniqueFileDescriptor always contains a valid FD (or is empty)